module/canvas/token-placement.mjs

```diff
@@ -0,0 +1,259 @@
+
+  /* -------------------------------------------- */
+
+  /**
+   * Pixel offset to ensure snapping occurs in middle of grid space.
+   * @param {PrototypeToken} token  Token for which to calculate the adjustment.
+   * @returns {{x: number, y: number}}
+   */
+  #getSnapAdjustment(token) {
+    const size = canvas.dimensions.size;
+    switch ( canvas.grid.type ) {
+      case CONST.GRID_TYPES.SQUARE:
+        return {
+          x: token.width % 2 === 0 ? Math.round(size * 0.5) : 0,
+          y: token.height % 2 === 0 ? Math.round(size * 0.5) : 0
+        };
+      default:
+        return { x: 0, y: 0 };
+    }
+  }
+
+  /* -------------------------------------------- */
+  /*  Event Handlers                              */
+  /* -------------------------------------------- */
+
+  /**
+   * Activate listeners for the placement preview.
+   * @returns {Promise}  A promise that resolves with the final placement if created.
+   */
+  #activatePreviewListeners() {
+    return new Promise((resolve, reject) => {
+      this.#events = {
+        cancel: this.#onCancelPlacement.bind(this),
+        confirm: this.#onConfirmPlacement.bind(this),
+        move: this.#onMovePlacement.bind(this),
+        resolve,
+        reject,
+        rotate: this.#onRotatePlacement.bind(this)
+      };
+
+      // Activate listeners
+      canvas.stage.on("mousemove", this.#events.move);
+      canvas.stage.on("mousedown", this.#events.confirm);
+      canvas.app.view.oncontextmenu = this.#events.cancel;
+      canvas.app.view.onwheel = this.#events.rotate;
+    });
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Shared code for when token placement ends by being confirmed or canceled.
+   * @param {Event} event  Triggering event that ended the placement.
+   */
+  async #finishPlacement(event) {
+    canvas.tokens._onDragLeftCancel(event);
+    canvas.stage.off("mousemove", this.#events.move);
+    canvas.stage.off("mousedown", this.#events.confirm);
+    canvas.app.view.oncontextmenu = null;
+    canvas.app.view.onwheel = null;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Move the token preview when the mouse moves.
+   * @param {Event} event  Triggering mouse event.
+   */
+  #onMovePlacement(event) {
+    event.stopPropagation();
+    if ( this.#throttle ) return;
+    const preview = this.#previews[0];
+    const adjustment = this.#getSnapAdjustment(preview);
+    const point = event.data.getLocalPosition(canvas.tokens);
+    const center = canvas.grid.getCenter(point.x - adjustment.x, point.y - adjustment.y);
+    preview.updateSource({
+      x: center[0] + adjustment.x - Math.round((this.config.tokens[0].width * canvas.dimensions.size) / 2),
+      y: center[1] + adjustment.y - Math.round((this.config.tokens[0].height * canvas.dimensions.size) / 2)
+    });
+    this.#placements[0].x = preview.x;
+    this.#placements[0].y = preview.y;
+    preview.object.refresh();
+    this.#throttle = true;
```

Comment:
Handling throttling via `requestAnimationFrame` is probably preferable I think. `#throttle` can become a simple boolean.

```suggestion
    if ( this.#throttle ) return;
    this.#throttle = true;
    const preview = this.#previews[0];
    const adjustment = this.#getSnapAdjustment(preview);
    const point = event.data.getLocalPosition(canvas.tokens);
    const center = canvas.grid.getCenter(point.x - adjustment.x, point.y - adjustment.y);
    preview.updateSource({
      x: center[0] + adjustment.x - Math.round((this.config.tokens[0].width * canvas.dimensions.size) / 2),
      y: center[1] + adjustment.y - Math.round((this.config.tokens[0].height * canvas.dimensions.size) / 2)
    });
    this.#placements[0].x = preview.x;
    this.#placements[0].y = preview.y;
    preview.object.refresh();
    requestAnimationFrame(() => this.#throttle = false);
```